README
This uses the Translation code from a defunct plugin at
https://sites.google.com/site/kashmirplugins/home/translator.

The YouTube trailer and search functionality has been removed (as it did
not work, and other YT plugins exist).

The code has been rationalized, and it should be (or will become)
Python3 compatible.

The entries here are:

src/
    The source files

MakeHTML5Entities.pl
    A script to create the HTML5Entities.py source from the full list of
    HTML5 entities at w3.org.

ipkg-build
    An edited version of the ipkg-build script from the Vix opkg-tools
    with a few modifications to make it a simple build script for this
    package.
    Just run it in this directory.

enigma2-plugin-extensions-translator_x.xx_all.ipk
    A package packed by ipkg-build.
    This will ONLY be packaged before a commit to Master
    it will NOT be updated for every Dev change made!!!

README
    This file.



CHANGES
Added files to the git repository.

========== 2.0 STARTS [21 Dec 2020]

Added build ipk package for 2.0

========== 2.0+ipk STARTS [21 Dec 2020]

README + Dev branch
    Added Dev branch and updated README to mention it

plugin.py
    Added ".decode('utf-8')" to urlopen() call. Needed for Py3, and
    missed on copying in from test code.
    Added plugin-source symlink for easier access to source files.

plugin.py
    Change the test for "no text" in translateText() and translateEPG()
    and correct the (previously wrong) indentation in the latter.

plugin.py
    Switch to using epgcache.lookupEvent() to get events (based on EPG
    TranslatorLite.
    Get the EPG starting from the normal histminutes previous, rather
    than just getting all EPG entries (which can go back well beyond
    that). Then start with the current event.

plugin.py
    Switch off wrapping at the beginning and end of the displayed list.

plugin.py
    Change the returns list for epgcache.lookupEvent() to be only those
    used, and parametrize the handling to robustify any additions or
    changes.

plugin.py
    Cache the EPG translations, so that we don't need to redo the web
    call on a return to a previously-seen item.

plugin.py
    Move the text translation call into a stand-alone DO_translation()
    function so that it may be called from anywhere.

plugin.py
    Add code to intercept EventView pop-ups and add translations there
    toggled by the Text key.

========== 2.01-rc1 STARTS [11 Jan 2021]


AutoflushCache.py
    Added code to handle cacheing the translations. In its own class.

Skin_medium.py, Skin_small.py
    (Re)added the label2 item for the Read, clear-cache button

plugin.py
    Change the User-Agent string to be "EPG Translator" based.
    Remember all events, not just a (configurable) number.
    Re-ordered some code.
    Added cacheing of translated text.
    Add configurable cacheing timeout.
    Add red-button handling to clear cache.
    Add handling of translations for EventView items using the Text
    button as a toggle.
    Removed My_EPGSel* functions.

*-2.01-rc1.ipk
    Built....

========== 2.01-rc2 STARTS [13 Jan 2021]

Skin_medium.py
Skin_small.py
    Enlarged the translatorConfig screen.
    Added the (forgotten) yellow image back to translatorMain.

plugin.py
    Add lang_flag() function, just to have only one place that knows how
    to form the pathnames.
    Don't add ch+/ (or rather "nextBouquet" and "prevBouquet") button
    bindings if we are viewing a recording, not </>.
    Use a ServiceEventTracker to call getEPG on service (channel)
    changes rather than doing this in zapUp/zapDown..  This means we
    don't try until the new channel is actually tuned it, and so works
    even when a timeshift is running (the user may be prompted before
    anything happens).
    Handle viewing a playback of a recording by creating a hand-crafted
    EPG with one entry.

README
    Update the release README to add more detail.

*-2.01-rc2.ipk
    Built 2.01-rc2.ipk.  2.01-rc1.ipk removed.

========== 2.01-rc3 STARTS

Skin_medium.py
Skin_small.py
    Fixed some minor separator geometry issues.

plugin.py
    Removed [..] trailing props before sending them for translation, and
    add them back afterwards for display.
    Prevents en -> de changing [S,AD] to [TRAURIG]

AutoflushCache.py
    Define, in __init__(), the value to return for "not found" key
    lookups. This allows the storage of (title, descr) tuples (and
    other things).

plugin.py
    Put list() around range()s, for Py3.
    Reformat language list (to use fewer lines).
    Update AutoflushCache() calls to store a (title, descr) tuple.
    Move 2000 char test into the DO_translation() function.
    Translate both the Title and Description for programmes (still done
    in a single call to translate.google.com)
    Replaced the isPlayback() test for being in playback mode, as it's
    Vix specific.
    Add a my_Sref() method, as we need a serviceref in a few locations.
    Add the current service (channel) name to the title in the
    translator EPG browser.
    strip() all text we are asked to translate.
    Append a newline to the begin text, and prepend one to the duration
    text instead.  Gives consistency.
    Cater for recordings with no Event info (happens for downloaded
    files).
    Add try/except clauses when fetching data to translate so that if
    something goes wrong we just fail to translate, rather than crash.
    Change the "EventID" for a recording to be the dev:inode for the
    recording, rather than its full pathname.
    In My_setEvent() get the title text from self["epg_eventname"], then
    update both that and setTitle() with the translation, as skins may
    be using either.

enigma2-plugin-extensions-translator_2.0-rc2_all.ipk
    Deleted
    Dev releases will no longer be put on github.
    They will be posted here (and later posts)
    https://www.world-of-satellite.com/showthread.php?31010-EPG-translator-(download-thread)&p=507698&viewfull=1#post507698
Added the 2.01-rc3 tag.

========== 2.01-rc4 STARTS

plugin.py
    Added a regex handler for [] groups being at the start of a
    descripton. Use it if char1 is [, use the end group match if chsar-1
    is ] and do not matching otherwise.
    Removed the 3-char limit for an individual tag within [].
    Updated version tag to 2.01-rc4

control
    Built ipk
    Added the 2.01-rc4 tag

========== 2.01-rc5 STARTS

plugin.py
    Add Mozilla/5.0 compat to User Agent string
    Version now shown in configuration menu.
    Raised encoded text length limit from 2000 to 7000 after testing
    what was allowed.
    Add "debug" print statements to some except clauses, so oddities
    can be recorded.
    Fixed My_setEvent() to use the new begin_matcher+end_matcher
    regexes.
    Updated version tag to 2.01-rc5

control
    Built ipk
    Added the 2.01-rc5 tag to git

========== 2.01-rc6 STARTS

plugin.py

MakeHTML5Entities.pl
#!/usr/bin/perl
#

# Create a name2codepoint dict for python using the *complete* list of
# HTML5 entities downloaded from the w3 org website.
#

use strict;
use warnings;

use LWP::Simple qw( get );

# Get the web page...
#
my $resp = get('https://dev.w3.org/html5/html-author/charref');
exit unless defined $resp;

# We get lines like this (wrapped here):
#   <tr title="U+00009 CHARACTER TABULATION" data-block="C0 Controls and Basic Latin"
#    data-category="Cc" data-set="mmlextra"><td class="character"> &#x00009;
#    <td class="named"><code>&amp;Tab;</code><td class="hex"><code>&amp;#x00009;
#    </code><td class="dec"><code>&amp;#9;</code><td class="desc">CHARACTER TABULATION
#
# The class="named" section may contain multiple, space-separated entity names.
# Creating a mapping of all entities for each hexcode
#
my %names4;
foreach my $ln (split "\n", $resp) {
    next unless $ln =~ /^<tr/;  # Skip if not a table row
# First we have to translate &amp; to &
#
    $ln =~ s/&amp;/&/g;
# Now extract the name(s) and hex code
#
    my ($nstr, $hexc) =
         ($ln =~ q%\"named\"><code>(.*?)</code>.*?\"hex\"><code>&#(.*?);%);
    $names4{$hexc} = $nstr;
}

# Print out the python file header
#
print <<'EOH';
# This is the full HTML5 Entity list from:
#   https://dev.w3.org/html5/html-author/charref
# This IS case sensitive!!
# The name2codepoint in htmlentitydefs is incomplete.
#
name2codepoint = {
EOH

# Now print out the name for the hex-code in ascending order
# We order them by hex code and handle any multiple entity names
# by writing multiple lines, one for each name.
#
my $nents = 0;
my $nchars = 0;
foreach my $hx (sort keys %names4) {
    $nchars++;
    foreach my $nm (split " ", $names4{$hx}) {
        $nents++;
        my $ent = substr($nm, 1, -1);
        printf "    %-34s : 0%s,\n", "'$ent'", $hx;
    }
}

# Now print the closing brace for the python file
#
print <<'EOF';
}
EOF

# Display stats on STDERR
#
print STDERR "Found $nents entries for $nchars chars\n";


src/usr/lib/enigma2/python/Plugins/Extensions/EPGTranslator/Skin_small.py
# Skin data for screenwidth <= 1280
# Only contains "class" data.  No instance data
#

class MySkinData:

# class translatorConfig bits
#
    translatorConfig_skin = """
        <screen position="center,center" size="811,525" backgroundColor="#20000000" title="EPG Translator Setup">
            <ePixmap position="133,0" size="545,50" pixmap="{plug_loc}/pic/translatorConfig.png" alphatest="blend" zPosition="1" />
            <ePixmap position="70,59" size="671,1" pixmap="{plug_loc}/pic/separator.png" alphatest="off" zPosition="1" />
            <widget name="config" position="10,60" size="658,182" itemHeight="38" scrollbarMode="showOnDemand" font="Regular;24" secondfont="Regular;24" zPosition="1" />
            <ePixmap position="70,244" size="671,1" pixmap="{plug_loc}/pic/separator.png" alphatest="off" zPosition="1" />
            <ePixmap position="210,254" size="18,18" pixmap="{plug_loc}/pic/buttons/green.png" alphatest="blend" zPosition="1" />
            <ePixmap position="525,254" size="18,18" pixmap="{plug_loc}/pic/buttons/red.png" alphatest="blend" zPosition="1" />
            <eLabel position="235,253" size="180,26" font="Regular;24" halign="left" text="Save" transparent="1" zPosition="1" />
            <eLabel position="550,253" size="180,26" font="Regular;24" halign="left" text="Cancel" transparent="1" zPosition="1" />
            <widget name="flag" position="261,250" size="288,288" alphatest="blend" zPosition="1" />
        </screen>
        """

# class translatorMain bits
#
#   "text" y-size is dynamic {size}
#
    translatorMain_skin = """
        <screen position="center,80" size="1000,610" title="EPG Translator">
            <ePixmap position="0,0" size="1000,50" pixmap="{plug_loc}/pic/translator.png" alphatest="blend" zPosition="1" />
            <ePixmap position="10,6" size="18,18" pixmap="{plug_loc}/pic/buttons/blue.png" alphatest="blend" zPosition="2" />
            <ePixmap position="10,26" size="18,18" pixmap="{plug_loc}/pic/buttons/yellow.png" alphatest="blend" zPosition="2" />
            <widget name="label" position="34,6" size="200,20" font="Regular;16" foregroundColor="black" backgroundColor="#FFFFFF" halign="left" transparent="1" zPosition="2" />
            <widget name="label2" position="34,26" size="200,20" font="Regular;16" foregroundColor="black" backgroundColor="#FFFFFF" halign="left" transparent="1" zPosition="2" />
            <widget render="Label" source="global.CurrentTime" position="740,0" size="240,50" font="Regular;24" foregroundColor="#697178" backgroundColor="#FFFFFF" halign="right" valign="center" zPosition="2">
                <convert type="ClockToText">Format:%H:%M:%S</convert>
            </widget>
            <widget name="timing" position="308,55" size="750,32" foregroundColor="orange" font="Bold;24" halign="left" zPosition="1" />
            <widget name="flag" position="10,90" size="288,288" alphatest="blend" zPosition="1" />
            <widget name="text" position="308,90" size="750,{size}" font="Regular;24" halign="left" zPosition="2" />
            <widget name="flag2" position="10,350" size="288,288" alphatest="blend" zPosition="1" />
            <widget name="text2" position="308,350" size="750,255" font="Regular;24" halign="left" zPosition="1" />
        </screen>
        """
    tMyes = "255"
    tMno  = "515"


MySD = MySkinData()


src/usr/lib/enigma2/python/Plugins/Extensions/EPGTranslator/AutoflushCache.py
# Implement a keyed-cache with elements that can be timed-out
# The time-out will be either the per-key or, if none given, a per-cache
# default relative value.
# Timeout flushes occur on add() or fetch() calls
#
import time

class AutoflushCache:
    def __init__(self, default_timeout):
        self.cache = {}         # key -> val pairs
        self.clear_time = []    # sorted (abs_timeout, key) tuples
        self.dflt_rel_timeout = abs(default_timeout)

# Change the default timeout.
# Only affects newly added keys (since we don't store how values
# were originally set)
#
    def change_timeout(self, new_timeout):
        self.dflt_rel_timeout = abs(new_timeout)

# A routine to remove any out-of-date items.
# These can be found at the head of the clear_time array.
# This returns the current time (for internal-use convenience)!
#
    def flush(self):
        now = time.time()
        clear_to = str(int(now))
        ix = 0
        while ix < len(self.clear_time):
            nx = ix + 1
            if self.clear_time[ix][0] > clear_to: break
            ix = nx
        if ix > 0:  # Something to clear
            for kx in range(ix):    # Delete the key/vals
                del self.cache[self.clear_time[kx][1]]
# Then delete the clear times
            del self.clear_time[0:ix]
        return now

# Add a key/val pair with an optional key-specific absolute timeout.
# This calls flush, whence it gets the current time.
# Is allowed to ovrwrite an exisiting key.
#
    def add(self, key, val, abs_timeout=0):
        now = self.flush()
# Iff this key already exists we need to remove its related clear_time
# entry.
        if key in self.cache:
            self.clear_time.remove(key)
        self.cache[key] = val
        if abs_timeout > 0:
            until = str(int(abs_timeout))
        else:
            until = str(int(now + self.dflt_rel_timeout))
        ix = 0
        while ix < len(self.clear_time):
            nx = ix + 1
            if self.clear_time[ix][0] > until: break
            ix = nx
        self.clear_time.insert(ix, (until, key))

# Get the value for a key.
# Calls flush() before fetching.
# Returns None if there is no such key.
#
    def fetch(self, key):
        self.flush()
        if key in self.cache:
            retval = self.cache[key]
        else:
            retval = None
        return retval

# Delete a key if present.
# Deleting a non-existent key is not an error
#
    def delete(self, key):
        if key in self.cache:
            del self.cache[key]
            self.clear_time.remove(key)

# Remove all keys
#
    def purge(self):
        self.cache = {}
        self.clear_time = []

# Destructor - probably what happens by default anyway...
#
    def __del__(self):
        self.purge()


src/usr/lib/enigma2/python/Plugins/Extensions/EPGTranslator/Skin_medium.py
# Skin data for screenwidth > 1280
# Only contains "class" data.  No instance data
#

class MySkinData:

# class translatorConfig bits
#
    translatorConfig_skin = """
        <screen position="center,center" size="1218,788" backgroundColor="#20000000" title="EPG Translator Setup">
            <ePixmap position="200,0" size="818,75" pixmap="{plug_loc}/pic/fhd-translatorConfig.png" alphatest="blend" zPosition="1" />
            <ePixmap position="110,89" size="998,1" pixmap="{plug_loc}/pic/separator.png" alphatest="off" zPosition="1" />
            <widget name="config" position="15,90" size="1188,268" itemHeight="57" scrollbarMode="showOnDemand" font="Regular;36" secondfont="Regular;36" zPosition="1" />
            <ePixmap position="110,369" size="998,1" pixmap="{plug_loc}/pic/separator.png" alphatest="off" zPosition="1" />
            <ePixmap position="318,384" size="27,27" pixmap="{plug_loc}/pic/buttons/fhd-green.png" alphatest="blend" zPosition="1" />
            <ePixmap position="795,384" size="27,27" pixmap="{plug_loc}/pic/buttons/fhd-red.png" alphatest="blend" zPosition="1" />
            <eLabel position="355,383" size="270,40" font="Regular;36" halign="left" text="Save" transparent="1" zPosition="1" />
            <eLabel position="833,383" size="270,40" font="Regular;36" halign="left" text="Cancel" transparent="1" zPosition="1" />
            <widget name="flag" position="464,450" size="288,288" alphatest="blend" zPosition="1" />
        </screen>
        """

# class translatorMain bits
#
#   "text" y-size is dynamic {size}
#
    translatorMain_skin = """
        <screen position="center,120" size="1500,915" title="EPG Translator">
            <ePixmap position="0,0" size="1500,75" pixmap="{plug_loc}/pic/fhd-translator.png" alphatest="blend" zPosition="1" />
            <ePixmap position="15,9" size="27,27" pixmap="{plug_loc}/pic/buttons/fhd-blue.png" alphatest="blend" zPosition="2" />
            <ePixmap position="15,39" size="27,27" pixmap="{plug_loc}/pic/buttons/fhd-yellow.png" alphatest="blend" zPosition="2" />
            <widget name="label" position="51,9" size="300,30" font="Regular;24" foregroundColor="black" backgroundColor="#FFFFFF" halign="left" transparent="1" zPosition="2" />
            <widget name="label2" position="51,39" size="300,30" font="Regular;24" foregroundColor="black" backgroundColor="#FFFFFF" halign="left" transparent="1" zPosition="2" />
            <widget render="Label" source="global.CurrentTime" position="1110,0" size="360,75" font="Regular;36" foregroundColor="#697178" backgroundColor="#FFFFFF" halign="right" valign="center" zPosition="2">
                <convert type="ClockToText">Format:%H:%M:%S</convert>
            </widget>
            <widget name="timing" position="318,90" size="1200,50" foregroundColor="orange" font="Bold;36" halign="left" zPosition="1" />
            <widget name="flag" position="15,140" size="288,288" alphatest="blend" zPosition="1" />
            <widget name="text" position="318,140" size="1200,{size}" font="Regular;36" halign="left" zPosition="2" />
            <widget name="flag2" position="15,530" size="288,288" alphatest="blend" zPosition="1" />
            <widget name="text2" position="318,530" size="1200,385" font="Regular;36" halign="left" zPosition="1" />
        </screen>
        """
    tMyes = "380"
    tMno  = "770"


MySD = MySkinData()
